### PR TITLE
[JENKINS-68947] Active Directory plugin fails on Java 17 with `IllegalAccessError`: "module `java.naming` does not export `com.sun.jndi.ldap` to unnamed module"

### PR DESCRIPTION
**Pull request by basil**
Tentative fix for [JENKINS-68947](https://issues.jenkins.io/browse/JENKINS-68947) pending testing. See self-review for an explanation of the changes.

### CODE REVIEW

File: .mvn/extensions.xml [hunk 1]
```diff
@@ -2,6 +2,6 @@
   <extension>
     <groupId>io.jenkins.tools.incrementals</groupId>
     <artifactId>git-changelist-maven-extension</artifactId>
-    <version>1.2</version>
+    <version>1.4</version>
```
Comment (basil) [inline .mvn/extensions.xml:5]: Unrelated code cleanup done while I was here: bumping the `git-changelist-maven-extension` to the latest version to facilitate producing incremental builds for testing purposes.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 2]
```diff
@@ -662,12 +662,13 @@ private LdapContext bind(String principalName, String password, SocketInfo serve
             Thread.currentThread().setName("Connecting to "+ldapUrl+" : "+oldName);
             LOGGER.fine("Connecting to " + ldapUrl);
             try {
+                props.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
```
Comment (basil) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java:665]: Consistent with e.g. spring-projects/spring-ldap@0a1d3f595b670412b7fd7ab739237c57381a7fa8. Rather than directly importing this class, which is supposed to be referenced only from the `java.naming` module, we pass it in as a string. The `java.naming` module can then interpret that string as it sees fit internally.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 3]
```diff
@@ -662,12 +662,13 @@ private LdapContext bind(String principalName, String password, SocketInfo serve
...
                 props.put(Context.PROVIDER_URL, ldapUrl);
                 props.put("java.naming.ldap.version", "3");
                 
                 customizeLdapProperties(props);
                 
-                LdapContext context = (LdapContext)LdapCtxFactory.getLdapCtxInstance(ldapUrl, props);
+                LdapContext context = new InitialLdapContext(props, null);
```
Comment (basil) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java:671]: Consistent with https://github.com/spring-projects/spring-ldap/blob/15aab1f8263a0048fe629facd388973e409b36fc/core/src/main/java/org/springframework/ldap/core/support/LdapContextSource.java.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 4]
```diff
@@ -700,7 +701,7 @@ private LdapContext bind(String principalName, String password, SocketInfo serve
                         // JENKINS-44787 It seems than to go back to plain-text LDAP does not work
                         // in all cases and to re-create the context is necessary
                         context.close();
-                        context = (LdapContext)LdapCtxFactory.getLdapCtxInstance(ldapUrl, props);
+                        context = new InitialLdapContext(props, null);
```
Comment (basil) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java:704]: Consistent with https://github.com/spring-projects/spring-ldap/blob/15aab1f8263a0048fe629facd388973e409b36fc/core/src/main/java/org/springframework/ldap/core/support/LdapContextSource.java.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 5]
```diff
@@ -734,7 +735,7 @@ private LdapContext bind(String principalName, String password, SocketInfo serve
         public DirContext createDNSLookupContext() throws NamingException {
             Hashtable env = new Hashtable();
             env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.dns.DnsContextFactory");
-            env.put("java.naming.provider.url", "dns:");
+            env.put(Context.PROVIDER_URL, "dns:");
```
Comment (basil) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java:738]: Unrelated code cleanup done while I was here: using a built-in constant where possible, consistent with the previous usage near line 666.

**Comment by basil** [comment]
Incremental build 2.27-rc704.a_d13ff057b_f7 is available for testing:

https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/active-directory/2.27-rc704.a_d13ff057b_f7/

For instructions on how to install an incremental build, see:

https://www.jenkins.io/doc/book/managing/plugins/#advanced-installation

**Review by olamy — APPROVED** [review]

**Comment by basil** [comment]
@famod @jimklimov Gentle ping. Were you interested in testing this change?

**Comment by jimklimov** [comment]
Sorry, did not notice the notification in my stream. Hope to give this a spin over the weekend.

**Comment by jimklimov** [comment]
Thanks, seems functional although threw one stack trace on start (posted in JIRA).

**Review by jimklimov — APPROVED** [review]

**Comment by basil** [comment]
Thanks @jimklimov! That exception does not look related to the changes in this PR. From my perspective, this PR is ready to merge and release.

**Comment by car-roll** [comment]
@basil https://github.com/jenkinsci/active-directory-plugin/releases/tag/active-directory-2.27

**Comment by basil** [comment]
Many thanks!
